File: src/main/java/com/github/unldenis/hologram/line/TextLine.java
package com.github.unldenis.hologram.line;


import com.github.unldenis.hologram.packet.PacketsFactory;
import com.github.unldenis.hologram.placeholder.Placeholders;
import com.github.unldenis.hologram.util.AABB;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.Nullable;

public final class TextLine implements ITextLine {

  private final Line line;
  private final Placeholders placeholders;
  private final boolean clickable;

  private String obj;
  private boolean isEmpty;
  private AABB hitbox;

  public TextLine(Line line, String obj, @Nullable Placeholders placeholders, boolean clickable) {
    this.line = line;
    this.placeholders = placeholders == null ? new Placeholders(0x00) : placeholders;
    this.clickable = clickable;

    this.obj = obj;
  }

  public TextLine(Line line, String obj, Placeholders placeholders) {
    this(line, obj, placeholders, false);
  }

  public TextLine(Line line, String obj) {
    this(line, obj, null);
  }

  @Override
  public boolean isClickable() {
    return clickable;
  }

  @Override
  public String parse(Player player) {
    return placeholders.parse(this.obj, player);
  }

  @Override
  public TextLine asTextLine() {
    return this;
  }

  @Override
  public Placeholders getPlaceholders() {
    return placeholders;
  }

  public AABB getHitbox() {
    return hitbox;
  }

  @Override
  public Plugin getPlugin() {
    return line.getPlugin();
  }

  @Override
  public Type getType() {
    return Type.TEXT_LINE;
  }

  @Override
  public int getEntityId() {
    return line.getEntityID();
  }

  @Override
  public Location getLocation() {
    return line.getLocation();
  }

  @Override
  public void setLocation(Location location) {
    line.setLocation(location);
    if (clickable) {
      double chars = this.obj.length();
      double size = 0.105;
      double dist = size * (chars / 2d);

      hitbox = new AABB(
          new AABB.Vec3D(-dist, -0.039, -dist),
          new AABB.Vec3D(dist, +0.039, dist));
      hitbox.translate(AABB.Vec3D.fromLocation(location.clone().add(0, 1.40, 0)));
    }
  }

  @Override
  public String getObj() {
    return this.obj;
  }

  @Override
  public void setObj(String obj) {
    this.obj = obj;
  }

  @Override
  public void hide(Player player) {
    line.destroy(player);
  }

  @Override
  public void teleport(Player player) {
    line.teleport(player);
  }

  @Override
  public void show(Player player) {
    isEmpty = this.obj.isEmpty();
    if(!isEmpty) {
      line.spawn(player);
      PacketsFactory.get()
          .metadataPacket(line.getEntityID(), parse(player), true, true, null)
          .send(player);
    }
  }

  @Override
  public void update(Player player) {
    byte spawnBefore = (byte) ((isEmpty ? 1 : 0) | (this.obj.isEmpty() ? 1 : 0) << 1);
    /*  0x00  = is already showed
        0x01  = is hided but now has changed
        0x02  = is already showed but is empty
        0x03  = is hided and isn't changed      */
    switch (spawnBefore) {
      case 0x03:
        break;
      case 0x02:
        line.destroy(player);
        isEmpty = true;
        break;
      case 0x01:
        line.spawn(player);
        isEmpty = false;
      case 0x00:
        PacketsFactory.get()
            .metadataPacket(line.getEntityID(), parse(player), spawnBefore == 0x01, spawnBefore == 0x01, null)
            .send(player);
    }
  }

}

File: src/main/java/com/github/unldenis/hologram/line/TextLine.kt
package com.github.unldenis.hologram.line

import com.github.unldenis.hologram.builder.interfaces.PlayerFun
import com.github.unldenis.hologram.ext.send
import com.github.unldenis.hologram.packet.IPacket
import com.github.unldenis.hologram.util.AABB
import com.github.unldenis.hologram.util.AABB.Vec3D
import org.bukkit.Location
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin

class TextLine(plugin: Plugin, obj: String, override val args: Array<PlayerFun>? = null,
    override val clickable: Boolean = false
) : ITextLine {

    private val line: Line = Line(plugin, EntityType.ARMOR_STAND)

    override var obj: String = ""

    var clickEvent : ClickEvent? = null

    init {

        if (args == null) {
            this.obj = obj
        } else {
            this.obj = obj.replace("{}", "%s")
        }
    }

    var hitbox: AABB? = null
        private set
    private var isEmpty = false

    override val textLine: TextLine
        get() = this

    override fun parse(player: Player): String {
        if (args == null) {
            return obj
        }
        val res = arrayOfNulls<Any>(args.size)
        for (i in args.indices) {
            res[i] = args[i](player)
        }

        return String.format(obj, args = res)
    }

    override fun onClick(clickEvent: ClickEvent) {
        this.clickEvent = clickEvent
    }

    override val plugin: Plugin
        get() = line.plugin
    override val type: ILine.Type
        get() = ILine.Type.TEXT_LINE
    override val entityId: Int
        get() = line.entityID
    override val location: Location?
        get() = line.location

    override fun setLocation(value: Location) {
        line.location = value
        if (clickable) {
            val chars = obj.length.toDouble()
            val size = 0.105
            val dist = size * (chars / 2.0)

            hitbox = AABB(
                Vec3D(-dist, -0.039, -dist),
                Vec3D(dist, +0.039, dist)
            ).also {
                it.translate(Vec3D.fromLocation(value.clone().add(0.0, 1.40, 0.0)))
            }
        }
    }

    override fun hide(player: Player) {
        line.destroy(player)
    }

    override fun teleport(player: Player) {
        line.teleport(player)
    }

    override fun show(player: Player) {
        isEmpty = obj.isEmpty()
        if (!isEmpty) {
            line.spawn(player)
            val packet = IPacket.get(IPacket.Type.METADATA_TEXT)
                .metadata(entityId, parse(player))
            packet.send(player)
        }
    }

    override fun update(player: Player) {
        val spawnBefore = ((if (isEmpty) 1 else 0) or ((if (obj.isEmpty()) 1 else 0) shl 1))
        /*  0x00  = is already showed
            0x01  = is hided but now has changed
            0x02  = is already showed but is empty
            0x03  = is hided and isn't changed      */
        when (spawnBefore) {
            0x03 -> {}
            0x02 -> {
                line.destroy(player)
                isEmpty = true
            }

            0x01 -> {
                line.spawn(player)
                isEmpty = false
                IPacket.get(IPacket.Type.METADATA_TEXT)
                    .metadata(entityId, parse(player)).send(player)
            }

            0x00 ->
                IPacket.get(IPacket.Type.METADATA_TEXT)
                    .metadata(entityId, parse(player), invisible = false)
                    .send(player)
        }
    }


}